library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/service/GafferPopNamedOperationService.java

```diff
@@ -44,44 +57,61 @@ public Type getType() {
-            for (final NamedOperationDetail namedOp : namedOps) {
-                if (namedOp.getOperationName().equals(name)) {
-                    if (namedOp.getOperationChainWithDefaultParams().getOutputClass().equals(Iterable.class)) {
-                        return (CloseableIterator<R>) CloseableIterator.of(graph.execute(new OperationChain.Builder()
-                                .first(new NamedOperation.Builder()
-                                        .name(name)
-                                        .build())
-                                .then(new GenerateObjects.Builder<GafferPopElement>()
-                                        .generator(new GafferPopElementGenerator(graph))
-                                        .build())
-                                .build()).iterator());
-                    } else {
-                        return CloseableIterator.of(Arrays.asList(graph.execute(new OperationChain.Builder()
-                                .first(new NamedOperation.Builder<Void, R>()
-                                        .name(name)
-                                        .build())
-                                .build())).iterator());
-                    }
-                }
-            }
-            throw new IllegalStateException("Named Operation not found");
+            return executeNamedOperation((String) params.get("execute"));
         } else if (params.containsKey("add")) {
-            final Map<String, Object> addParams = (Map) params.get("add");
-            graph.execute(new OperationChain.Builder()
-                    .first(new AddNamedOperation.Builder()
-                            .name((String) addParams.get("name"))
-                            .operationChain((String) addParams.get("opChain"))
-                            .build())
-                    .build());
-            return CloseableIterator.empty();
+            return addNamedOperation((Map<String, String>) params.get("add"));
         } else {
             throw new IllegalStateException("Missing parameter");
         }
     }
 
+    protected CloseableIterator<R> executeNamedOperation(final String name) {
+        // Fetch details for the requested Named Operation or throw an exception if it's not found
+        Iterable<NamedOperationDetail> allNamedOps = graph.execute(new OperationChain.Builder().first(new GetAllNamedOperations()).build());
+        NamedOperationDetail namedOperation = StreamSupport.stream(allNamedOps.spliterator(), false)
+                .filter((NamedOperationDetail nd) -> nd.getOperationName().equals(name)).findFirst()
+                .orElseThrow(() -> new IllegalStateException("Named Operation not found"));
+
+        // If the Named Operation outputs an Iterable, then execute and convert Elements to GafferPopElement
+        // Else execute and wrap the output as an iterator
+        if (namedOperation.getOperationChainWithDefaultParams().getOutputClass().equals(Iterable.class)) {
```

Comment:
This doesn't confirm that the Iterable return will be of Elements, only that it is an Iterable.

The key is how to tell **before** the results are returned. This is the closest we can do. I don't think the Named Operation stores the type of Object that is inside the Iterable anyway, so we can't get this information.

With the results, it's easy to tell if the results of an Operation are an Iterable. To tell if the Iterable has Elements is more difficult, but can be done. However, we need to know before we get any results in order to run the GafferPop Element generator. We can't get the results and somehow put them back in to run the generator. It might be possible to conditionally run the generator within the Gaffer Operation? Or we could create a wrapped Iterable with the generator to be able to do this after getting the results.

Reply:
If it can't be determined before the operation is executed then I agree this is the best we can do, as I think we want the generator to run inside Gaffer.